src/distancematrix.rs

```diff
@@ -0,0 +1,220 @@
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixType {
+    Square,
+    Lower,
+    Upper,
+}
+
+/// A distance matrix containing a list of taxon names and a matrix of pairwise distances.
+#[derive(Debug)]
+pub struct DistanceMatrix {
+    pub names: Vec<String>,
+    pub distances: Vec<Vec<f32>>,
+    pub matrix_type: MatrixType,
+}
+
+#[derive(Error, Debug)]
+pub enum DistanceMatrixError {
+    #[error("names and matrix do not have matching length")]
+    LengthError,
+    #[error("matrix has unrecognized shape")]
+    ShapeError,
+}
+
+impl DistanceMatrix {
+    /// The number of sequences (rows) in the matrix.
+    pub fn len(&self) -> usize {
+        self.names.len()
+    }
+
+    /// Create a new DistanceMatrix.
+    ///
+    /// `distances` must be a matrix with the same number of rows as `names`, and follow one of the accepted shapes.
+    pub fn new(names: Vec<String>, distances: Vec<Vec<f32>>) -> Result<Self, DistanceMatrixError> {
+        if names.len() != distances.len() {
+            Err(DistanceMatrixError::LengthError)
+        } else {
+            let n = names.len();
+            let square = distances.iter().all(|x| x.len() == n);
+            let lower = distances.iter().enumerate().all(|(i, x)| x.len() == i);
+            let upper = distances
+                .iter()
+                .enumerate()
+                .all(|(i, x)| x.len() == n - 1 - i);
+            let matrix_type = match (square, lower, upper) {
+                (true, _, _) => MatrixType::Square,
+                (_, true, _) => MatrixType::Lower,
+                (_, _, true) => MatrixType::Upper,
+                _ => Err(DistanceMatrixError::ShapeError)?,
+            };
+            Ok(Self {
+                names,
+                distances,
+                matrix_type,
+            })
+        }
+    }
+
+    // Converts an index (i,j) to the right index to use given the shape of the matrix.
+    fn get_index(&self, (i, j): (usize, usize)) -> (usize, usize) {
+        match self.matrix_type {
+            MatrixType::Square => (i, j),
+            MatrixType::Lower => {
+                assert!(i != j);
+                (max(i, j), min(i, j))
+            }
+            MatrixType::Upper => {
+                assert!(i != j);
+                (min(i, j), max(i, j) - min(i, j) - 1)
+            }
+        }
+    }
+}
+
+/// Index access into the DistanceMatrix, taking into account the shape.
+///
+/// Indices should be used as if the matrix was square. For triangular matrices,
+/// `i` and `j` must be distinct, and `(i,j)` and `(j,i)` represent the same
+/// element.
+///
+/// # Example
+///
+/// ```
+/// use bio_types::distancematrix::DistanceMatrix;
+///
+/// let mut t = DistanceMatrix::new(vec!["a".to_string(), "b".to_string()],
+/// vec![
+///  vec![],
+///  vec![1.0],
+/// ]).unwrap();
+/// assert_eq!(t[(1, 0)], 1.0);
+/// t[(0, 1)] = 2.0;
+/// assert_eq!(t[(1, 0)], 2.0);
+/// ```
+impl Index<(usize, usize)> for DistanceMatrix {
```

Comment:
Instead of this (or in addition to this), it would also be possible to overload `Index<usize>` to return a helper type that contains a reference to the `DistanceMatrix` and the first argument, and only resolves the actual index after indexing the helper type.